### PR TITLE
fix(mito2): adapt batch size for wide rows

### LINKED ISSUE #8479: Unable to flush table if batched strings overflow the string array
### What type of bug is this?

Unexpected error

### What subsystems are affected?

Standalone mode, Storage Engine, Datanode

### Minimal reproduce step

Ingest a table with strings that longer than 256KB.

### What did you expect to see?

Ingestion rate is stable

### What did you see instead?

Write buffer full error.

### What operating system did you use?

Unrelated

### What version of GreptimeDB did you use?

v1.1.2

### Relevant log output and stack trace

```bash
Failed to flush region xxxx(xxxx, 1)
Failed to compute arrow arrays OffsetOverflowError(2147721975)
```

### PR DESCRIPTION
**Pull request by evenyag**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)

Closes #8479.

## What's changed and what's your intention?

This PR adapts Mito's execution batch size to the estimated decoded row width so that batches target approximately 64 MiB instead of always using the default row count. This prevents wide string rows from overflowing Arrow string offsets during flush and compaction.

The estimated batch size is propagated through memtable flushes, bulk-part compaction, region compaction scans, and Parquet readers. Bulk memtable merge groups are also bounded by the configured compaction output size so that already-large parts are not merged into oversized intermediate batches.

Unit tests cover batch-size estimation, size-aware part grouping, and compaction scan batch-size selection. No additional validation was run while creating this PR.

## PR Checklist
Please convert it to a draft if some of the following conditions are not met.

- [x] I have written the necessary rustdoc comments.
- [x] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.
- [x] API changes are backward compatible.
- [x] Schema or data changes are backward compatible.

### CODE REVIEW

**Comment by evenyag** [comment]
@codex review

File: src/mito2/src/memtable/bulk.rs [hunk 1]
```diff
@@ -497,6 +546,7 @@ impl Memtable for BulkMemtable {
             predicate.predicate().cloned(),
             options.for_flush,
             options.pre_filter_mode,
+            options.batch_size,
```
[the remark this answers is not in the document]
Comment (evenyag) [reply src/mito2/src/memtable/bulk.rs:517]: We intentionally keep existing raw Arrow RecordBatch boundaries here. Such a batch has already been constructed successfully, so its Utf8/Binary offsets are valid; slicing it does not make that batch safer. The adaptive batch size is used when readers or merge operators assemble new batches, which is where offset growth can occur. I also added a test that a raw bulk iterator preserves existing batch boundaries even when the context hint is smaller.

**Comment by evenyag** [comment]
@codex review

File: src/mito2/src/memtable/bulk.rs [hunk 2]
```diff
@@ -1183,11 +1204,16 @@ impl MemtableCompactor {
             .max()
             .unwrap_or(0);
 
+        let batch_size = crate::batch_size::estimate_batch_size([(
+            estimated_total_rows as u64,
+            estimated_total_bytes as u64,
+        )]);
```
[the remark this answers is not in the document]
Comment (evenyag) [reply src/mito2/src/memtable/bulk.rs]: Addressed in c383910a70. Encoded bulk parts now derive adaptive batch sizing from the Parquet row group with the largest uncompressed size, paired with that same row group row count. Compressed bytes remain unchanged for merge and encoding thresholds. Added a focused multi-row-group regression test.

**Comment by evenyag** [comment]
@codex review

File: src/mito2/src/memtable/bulk.rs [hunk 3]
```diff
@@ -1009,6 +1029,32 @@ impl PartToMerge {
...
+                .iter()
+                .map(|row_group| {
+                    let uncompressed_bytes = row_group
+                        .columns()
+                        .iter()
+                        .map(|column| column.uncompressed_size() as u64)
+                        .sum();
+                    (row_group.num_rows() as u64, uncompressed_bytes)
+                })
+                .max_by_key(|(_, uncompressed_bytes)| *uncompressed_bytes),
```
[the remark this answers is not in the document]
Comment (evenyag) [reply src/mito2/src/memtable/bulk.rs:1054]: The risk is low.

**Review by waynexia — APPROVED** [review]
